clippy_lints/src/methods/unnecessary_fold.rs

```diff
@@ -105,12 +113,47 @@ fn check_fold_with_op(
             sugg,
             applicability,
         );
+        return true;
     }
+    false
 }
 
-pub(super) fn check(
+fn check_fold_with_method(
     cx: &LateContext<'_>,
     expr: &hir::Expr<'_>,
+    acc: &hir::Expr<'_>,
+    fold_span: Span,
+    method: Symbol,
+    replacement: Replacement,
+) {
+    // Extract the name of the function passed to `fold`
+    if let Res::Def(DefKind::AssocFn, fn_did) = acc.res_if_named(cx, method)
+        // Check if the function belongs to the operator
+        && cx.tcx.is_diagnostic_item(method, fn_did)
```

Comment:
You have to check that the iterator's `Item` type implements a compatible `Sum` or `Product` trait.